catalyst_voices_packages/catalyst_cardano_serialization/lib/src/fees.dart

```diff
@@ -1,35 +1,165 @@
+  ///
+  /// > This formula does not account for smart contract scripts.
+  int linearFee(int bytesCount) => bytesCount * coefficient + constant;
+
+  /// Calculates the fee for a transaction using the tiered pricing model.
+  /// This includes both the linear fee and, if applicable, a reference script
+  /// fee.
+  int tieredFee(int txBytes, int refScriptsBytes) {
+    final txFee = linearFee(txBytes);
+    final scriptFee = refScriptByteCost > 0 && refScriptsBytes > 0
+        ? refScriptFee(
+            multiplier,
+            sizeIncrement,
+            refScriptByteCost,
+            refScriptsBytes,
+          )
+        : 0;
+    return txFee + scriptFee;
+  }
+
+  /// Calculates the fee for the reference scripts.
   ///
-  /// The formula doesn't take into account smart contract scripts.
-  Coin minNoScriptFee(Transaction tx) {
-    final bytesCount = cbor.encode(tx.toCbor()).length;
-    return Coin(bytesCount) * coefficient + constant;
+  /// The reference script fee is based on a tiered pricing model:
+  /// - The fee increases for each additional `25KiB` chunk of reference
+  ///   script size.
+  /// - The first `25KiB` is charged at a base cost per byte, and subsequent
+  ///   chunks are charged at progressively higher rates, multiplied by `1.2`.
+  int refScriptFee(
+    double multiplier,
+    int sizeIncrement,
+    int refScriptCostPerByte,
+    int resScriptSize,
+  ) {
+    int calcRefScriptFee(double acc, double curTierPrice, int n) =>
+        n <= sizeIncrement
+            ? (acc + n * curTierPrice).floor()
+            : calcRefScriptFee(
+                acc + curTierPrice * sizeIncrement,
+                curTierPrice * multiplier,
+                n - sizeIncrement,
+              );
+    return calcRefScriptFee(
+      0,
+      refScriptCostPerByte.toDouble(),
+      resScriptSize,
+    );
+  }
+
+  /// Calculates the total size of reference scripts used in a transaction.
+  ///
+  /// This includes the sizes of reference scripts from both inputs and
+  /// reference inputs. Outputs without reference scripts do not contribute
+  /// to the total size.
+  ///
+  /// Duplicate reference scripts are counted each time they appear in the
+  /// transaction, i.e., when the same script is used on different inputs.
+  ///
+  /// However, any input that appears in both regular inputs and
+  /// reference inputs of a transaction is only used once in this computation.
+  int _calculateReferenceScriptSize(Set<TransactionUnspentOutput> inputs) {
+    final totalSize = inputs.fold<int>(
+      0,
+      (prevSize, input) =>
+          prevSize +
+          switch (input.output) {
+            final TransactionOutput output =>
+              output.scriptRef != null ? output.scriptRef!.length : 0,
+            _ => 0,
+          },
+    );
+
+    return totalSize;
   }
 
   @override
-  List<Object?> get props => [constant, coefficient];
+  List<Object?> get props => [constant, coefficient, refScriptByteCost];
```

Comment:
Please add all missing fields: multiplier, sizeIncrement, maxRefScriptSize.

The `props` are used for hashing/data equality.

Reply:
Thank you for pointing that out. 
Initially, I added those fields as static constants, but later decided to include them as constructor parameters, but I missed updating the props list.